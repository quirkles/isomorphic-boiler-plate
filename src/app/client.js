/* globals document, window */

import React from 'react';
import { createStore, applyMiddleware, compose } from 'redux';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import { routerMiddleware } from 'react-router-redux';
import { renderRoutes } from 'react-router-config';
import createBrowserHistory from 'history/createBrowserHistory';

import routes from './routes';
import reducers from './reducers';

const history = createBrowserHistory();

const middleware = routerMiddleware(history);

const rootEl = document.getElementById('app');

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(
  reducers,
  composeEnhancers(applyMiddleware(middleware)),
);

/*
configureStore takes 3 args, the root reducer, init store state, and additional middlewares
by default, configureStore comes with redux-logger middleware. Here we
registering our fetchMiddleware
*/

export const getDomTree = () => (
  <Provider store={store}>
    <Router history={history}>{renderRoutes(routes)}</Router>
  </Provider>
);

ReactDOM.hydrate(getDomTree(), rootEl);
