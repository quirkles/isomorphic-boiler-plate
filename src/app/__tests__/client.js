import ReactDOM from 'react-dom';

import { getDomTree } from '../client';

jest.mock('react-dom', () => ({
  hydrate: jest.fn(),
}));

jest.mock('../routes', () => 'ROUTES');
jest.mock('../reducers', () => 'REDUCERS');

describe('client bootstrap', () => {
  it('renders the dom tree and hydrates the initial state', () => {
    const renderedDomTree = getDomTree();
    expect(renderedDomTree).toMatchSnapshot();
    expect(ReactDOM.hydrate).toHaveBeenCalledWith(
      renderedDomTree,
      'element with id: app',
    );
  });
});
