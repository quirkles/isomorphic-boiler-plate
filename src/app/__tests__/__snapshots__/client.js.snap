// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client bootstrap renders the dom tree and hydrates the initial state 1`] = `
<Provider
  store="{\\"reducers\\":\\"REDUCERS\\",\\"middleware\\":\\"\\\\\\"routerMiddleware with \\\\\\\\\\\\\\"BROWSER_HISTORY\\\\\\\\\\\\\\"\\\\\\"\\"}"
>
  <Router
    history="BROWSER_HISTORY"
  >
    <div>
      "ROUTES"
    </div>
  </Router>
</Provider>
`;
